src/backend/distributed/utils/maintenanced.c

```diff
@@ -292,45 +336,101 @@ CitusMaintenanceDaemonMain(Datum main_arg)
 	BackgroundWorkerHandle *backgroundTasksQueueBgwHandle = NULL;
 	bool backgroundTasksQueueWarnedForLock = false;
 
-	/*
-	 * We do metadata sync in a separate background worker. We need its
-	 * handle to be able to check its status.
-	 */
-	BackgroundWorkerHandle *metadataSyncBgwHandle = NULL;
 
-	/*
-	 * Look up this worker's configuration.
-	 */
-	LWLockAcquire(&MaintenanceDaemonControl->lock, LW_EXCLUSIVE);
+	MaintenanceDaemonDBData *myDbData = NULL;
 
-	MaintenanceDaemonDBData *myDbData = (MaintenanceDaemonDBData *)
-										hash_search(MaintenanceDaemonDBHash, &databaseOid,
-													HASH_FIND, NULL);
-	if (!myDbData)
+	bool isMainDb = false;
+
+	if (databaseOid == 0)
 	{
-		/*
-		 * When the database crashes, background workers are restarted, but
-		 * the state in shared memory is lost. In that case, we exit and
-		 * wait for a session to call InitializeMaintenanceDaemonBackend
-		 * to properly add it to the hash.
-		 */
+		char *databaseName = MainDb;
+		isMainDb = true;
 
-		proc_exit(0);
-	}
+		BackgroundWorkerInitializeConnection(databaseName, NULL, 0);
+
+		/* Now we have a valid MyDatabaseId. */
+		/* Insert the daemon instance to the hash table. */
+		bool found;
+
+		LWLockAcquire(&MaintenanceDaemonControl->lock, LW_EXCLUSIVE);
+
+		myDbData = (MaintenanceDaemonDBData *) hash_search(
+			MaintenanceDaemonDBHash,
+			&MyDatabaseId,
+			HASH_ENTER_NULL,
+			&found);
 
-	if (myDbData->workerPid != 0)
+		if (!myDbData)
+		{
+			/*
+			 * When the database crashes, background workers are restarted, but
+			 * the state in shared memory is lost. In that case, we exit and wait
+			 * for Postmaster calling __PG_Init which in turn calls
+			 * InitializeMaintenanceDaemonForAdminDB.
+			 */
```

Comment:
I actually think it might be good to add this insertion logic (including zero-ing) to a function that is called here and by `InitializeMaintenanceDaemonBackend`.

Reply:
I think my main source of finding this code confusing is the two branches (`if (databaseOid == 0)` and `else`) that seem similar, but are actually quite different. I think it would help readability to split these a bit better. Either by putting their bodies in a function, or creating a new `CitusMaintenanceDaemonMain` function for the databaseOid == 0 case. These could then call some functions for the shared code. But I think having these long branches but quite different branches make the code hard to follow, since now there's two paths that you need to look at carefully.